File: Coins_in_a_Line_III.cpp
class Solution {
public:
    /**
     * @param values: a vector of integers
     * @return: a boolean which equals to true if the first player will win
     */
    bool firstWillWin(vector<int> &values) {
        // write your code here
        // read from:http://articles.leetcode.com/2011/02/coins-in-line.html
        if (values.size() % 2 == 0)
            return firstWillWinEven(values);
        int sum = 0;
        for (auto& val : values) sum += val;
        vector<vector<int>> table(3, vector<int>(values.size(), 0));
        //table[n][i] denotes the most sums of numbers which are from ith number to (i+n) th number in values
        for (int n = 0; n < values.size(); ++n){
            for (int i = 0; i + n < values.size(); ++i){
                int a = i + 2 < values.size() && n - 2 > -1 ? table[(n-2)%3][i+2] : 0;
                int b = i + 1 < values.size() && n - 2 > -1 ? table[(n-2)%3][i+1] : 0;
                int c = n - 2 > -1 ? table[(n-2)%3][i] : 0;
                table[n%3][i] = max(values[i] + min(a,b), values[i + n] + min(b, c));
            }
        }
        return table[(values.size()-1)%3][0] > sum - table[(values.size()-1)%3][0];
    }
    bool firstWillWinEven(vector<int>& values){
        int even = 0;
        int odd = 0;
        for (int i = 0; i < values.size(); ++i){
            if (i % 2 == 0) even += values[i];
            else odd += values[i];
        }
        return odd != even;
    }

};

File: Coins in a Line III.py
class Solution:
    # @param values: a list of integers
    # @return: a boolean which equals to True if the first player will win
    def firstWillWin(self, values):
        # write your code here
        
        def memorySearch(start, end, dp, values):
            if start > end or 0 > start or start >= length or end < 0 or end >= length:
                return 0
            if dp[start][end] > 0:
                return dp[start][end]
            
            dp[start][end] = max(
                min(memorySearch(start + 2, end, dp, values), memorySearch(start + 1, end - 1, dp, values)) + values[start], 
                min(memorySearch(start, end - 2, dp, values), memorySearch(start + 1, end - 1, dp, values)) + values[end])
            return dp[start][end]
        
        length = len(values)
        dp = [[0] * length for __ in range(length)]
        return memorySearch(0, length - 1, dp, values) * 2 > sum(values)